internal/app/machined/pkg/runtime/v1alpha1/platform/hcloud/metadata.go

```diff
@@ -63,37 +52,44 @@ type NetworkConfig struct {
 	} `yaml:"config"`
 }
 
-func (h *Hcloud) getMetadata(ctx context.Context) (*MetadataConfig, error) {
-	log.Printf("fetching hostname from: %q", HCloudHostnameEndpoint)
+func (h *Hcloud) getMetadata(ctx context.Context) (metadata *MetadataConfig, err error) {
+	getMetadataKey := func(key string) (string, error) {
+		res, metaerr := download.Download(ctx, fmt.Sprintf("%s/%s", HCloudMetadataEndpoint, key),
+			download.WithErrorOnNotFound(errors.ErrNoConfigSource),
+			download.WithErrorOnEmptyResponse(errors.ErrNoConfigSource))
+		if metaerr != nil && !stderrors.Is(metaerr, errors.ErrNoConfigSource) {
+			return "", fmt.Errorf("failed to fetch %q from IMDS: %w", key, metaerr)
+		}
+
+		return string(res), nil
+	}
+
+	metadata = &MetadataConfig{}
 
-	host, err := download.Download(ctx, HCloudHostnameEndpoint,
-		download.WithErrorOnNotFound(errors.ErrNoHostname),
-		download.WithErrorOnEmptyResponse(errors.ErrNoHostname))
-	if err != nil && !stderrors.Is(err, errors.ErrNoHostname) {
+	if metadata.Hostname, err = getMetadataKey("hostname"); err != nil {
 		return nil, err
 	}
 
-	log.Printf("fetching instance-id from: %q", HCloudInstanceIDEndpoint)
+	if metadata.InstanceID, err = getMetadataKey("instance-id"); err != nil {
+		return nil, err
+	}
 
-	instanceID, err := download.Download(ctx, HCloudInstanceIDEndpoint,
-		download.WithErrorOnNotFound(errors.ErrNoHostname),
-		download.WithErrorOnEmptyResponse(errors.ErrNoHostname))
-	if err != nil && !stderrors.Is(err, errors.ErrNoHostname) {
+	if metadata.AvailabilityZone, err = getMetadataKey("availability-zone"); err != nil {
 		return nil, err
 	}
 
-	log.Printf("fetching externalIP from: %q", HCloudExternalIPEndpoint)
+	// Origianl CCM/CSI uses first part of availability-zone to define region name.
+	// But metadata has different value.
+	// We will follow official behavior.
+	metadata.Region = strings.SplitN(metadata.AvailabilityZone, "-", 2)[0]
+
+	// if metadata.Region, err = getMetadataKey("region"); err != nil {
```

Comment:
should be removed?

Reply:
I was confused a little bit with it. 
I almost get used to using the region names like eu-central/us-east. 
But csi does not work well with that code names. I commented on it for the future. Maybe it helps to understand why we do not use the region value from the meta server.

Reply:
we prefer not to have commented out code... we can leave some comment explaining the intent, but not the code

Reply:
gotcha 👍